rig-neo4j/src/vector_index.rs

```diff
@@ -0,0 +1,358 @@
+
+    pub fn new(
+        graph: Graph,
+        embedding_model: M,
+        index_config: IndexConfig,
+        search_params: SearchParams,
+    ) -> Self {
+        Self {
+            graph,
+            embedding_model,
+            index_config,
+            search_params,
+        }
+    }
+
+    /// Calls the `CREATE VECTOR INDEX` Neo4j query and waits for the index to be created.
+    /// A newly created index is not immediately fully available but is created (ie data is indexed) in the background.
+    ///
+    /// ❗ If there is already an index targetting the same node label and property, the new index creation will fail.
+    ///
+    /// ### Arguments
+    /// * `node_label` - The label of the nodes to which the index will be applied. For example, if your nodes have
+    ///                  the label `:Movie`, pass "Movie" as the node_label parameter.
+    /// * `embedding_prop_name`(optional) - The name of the property that contains the embedding vectors. Defaults to "embedding".
+    ///
+    pub async fn create_and_await_vector_index(
+        &self,
+        node_label: String,
+        embedding_prop_name: Option<String>,
+    ) -> Result<(), VectorStoreError> {
+        // Create a vector index on our vector store
+        tracing::info!("Creating vector index {} ...", self.index_config.index_name);
+
+        let property = embedding_prop_name.unwrap_or("embedding".to_string());
+        let create_vector_index_query = format!(
+            "
+            CREATE VECTOR INDEX $index_name IF NOT EXISTS
+            FOR (m:{})
+            ON m.{}
+            OPTIONS {{
+                indexConfig: {{
+                    `vector.dimensions`: $dimensions,
+                    `vector.similarity_function`: $similarity_function
+                }}
+            }}",
+            node_label, property
+        );
+
+        self.graph
+            .run(
+                neo4rs::query(&create_vector_index_query)
+                    .param("index_name", self.index_config.index_name.clone())
+                    .param(
+                        "similarity_function",
+                        self.index_config.similarity_function.clone().to_bolt_type(),
+                    )
+                    .param("dimensions", self.embedding_model.ndims() as i64),
+            )
+            .await
+            .map_err(|e| VectorStoreError::DatastoreError(Box::new(e)))?;
+
+        // Check if the index exists with db.awaitIndex(), the call timeouts if the index is not ready
+        let index_exists = self
+            .graph
+            .run(
+                neo4rs::query("CALL db.awaitIndex($index_name, 10000)")
+                    .param("index_name", self.index_config.index_name.clone()),
+            )
+            .await;
+
+        if index_exists.is_err() {
+            tracing::warn!(
+                "Index with name `{}` is not ready or could not be created.",
+                self.index_config.index_name
+            );
```

Comment:
Shouldn't this also return an `Error`? Cause right now the function returns `Ok` even if the index was not created

Reply:
Good question. I was not able to parse the error coming from this call. Partly because if the creation fails (for instance if there is already an index of another name) the DB does not fire an error.

![image](https://github.com/user-attachments/assets/84c0c458-15d1-4213-8903-f1f3146438d1)